### ISSUE TITLE
[FP] Array and string offset validity

### ISSUE THREAD

**Issue by mishavantol** [kalessil/phpinspectionsea#618]
```php
<?php
$arr = [1, 2, 3];
$foo = array_slice($arr, 0, 2);
$bar = $foo[0];

$foo = array_slice([1, 2, 3], 0, 2);
$bar = $foo[0];

$foo = array_slice(array_reverse($arr), 0, 2);
$bar = $foo[0];
```

I'm experiencing a FP on '$foo[0]' on line 7 and 10. (but not on line 4)

> '$foo' may not support offset operations (or it's type not annotated properly: [0, array]).

**Comment by kalessil** [comment]
@mishavantol : let me guess, that's an EAP?

**Comment by b1rdex** [comment]
I'm on last EAP, can't reproduce. No warnings in this code at all.

Could you please check types reported of `$foo` and `array_slice` with ctrl+q?

**Comment by mishavantol** [comment]
@kalessil I'm usually working on EAP, but the same thing happens on 2017.2.4 too.

I've switched to 2017.2.4, disabled all custom plugins and tried again. See the attached screendump for the results. 

@b1rdex ctrl-q doesn't do anything on the mac, with ctrl-space I can see the types, all `array` here too.

<img width="936" alt="screen shot 2017-11-02 at 20 50 25" src="https://user-images.githubusercontent.com/2063695/32346967-846d9100-c00f-11e7-9fed-3a764d1174a2.png">

**Comment by b1rdex** [comment]
@mishavantol I mead `quick documentation` (it's in view menu).

**Comment by mishavantol** [comment]
@b1rdex 
<img width="589" alt="screen shot 2017-11-03 at 08 44 59" src="https://user-images.githubusercontent.com/2063695/32364167-9e00d172-c073-11e7-98ad-f59149855c64.png">

On `array_slice` I get the documentation for the function.

**Comment by b1rdex** [comment]
@mishavantol is return type of `array_slice` equals `0|array` too? Looks like 3rd party issue, do you have other plugins enabled?

**Comment by mishavantol** [comment]
@b1rdex the return type of `array_slice` is the expected `array`. The first screenshot was phpstorm with all custom plugins disabled. But looking at the return type this is most likely an phpstorm issue. When I change it to `array_slice([1, 2, 3], 12)` the return type becomes `12|array`. The second parameter seem to end up in the return type.

I'll create an issue in the phpstorm issue tracker.

**Comment by b1rdex** [comment]
Funny thing: I was going to ask if you have custom phpstorm stubs attached, because there are specifyied return types for `array_*` functions (see https://github.com/JetBrains/phpstorm-stubs/blob/master/meta/.phpstorm.meta.php). Then I realised that this "module" is turned off in settings, becuase I used `think modules to installed in interpreter` (this should be reported to jetbrains too, this module must not be disabled). I turned this on and now your issue is reproducible 😁 . Reopen please and provide a link to jetbrains issue you opened.

**Comment by b1rdex** [comment]
I created issue with this problem — https://youtrack.jetbrains.com/issue/WI-38649.

**Comment by mishavantol** [comment]
👍

**Comment by kalessil** [comment]
Closing issue. Thank you for back-linking!
